src/utils/SubscribableRequestQueue.ts

```diff
@@ -0,0 +1,171 @@
+import RequestQueue from "./RequestQueue";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Resolver = (value?: any) => void;
+type Rejecter = (reason?: unknown) => void;
+
+type RequestSubscription = {
+  subscriberId: number;
+  resolve: Resolver;
+  reject: Rejecter;
+};
+
+/**
+ * An extension of `RequestQueue` that adds a concept of "subscribers," which may share references to a single request
+ * or cancel their subscription without disrupting the request for other subscribers.
+ */
+export default class SubscribableRequestQueue {
+  private queue: RequestQueue;
+
+  /** Number of subscribers, used for generating unique subscriber IDs. */
+  private subscriberIds: number;
```

Comment:
Nit: `numSubscribers` might be clearer, or `numSubscriberIds`? The comment is useful but when I see this variable name I expect it to be an array of subscriber IDs or something.